discovery.rst

```diff
@@ -123,53 +148,89 @@ This type of discovery finds a URI factory for a PSR-7_ URI implementation::
 Integrating your own implementation with the discovery mechanism
 ----------------------------------------------------------------
 
-The ``php-http/discovery`` has built-in support for some implementations.
-To use a different implementation or override the default when several implementations are available in your code base,
-you can register a class explicitly with the corresponding discovery service. For example::
-
-    HttpClientDiscovery::register('Acme\MyClient', true);
-
-- ``class``: The class that is instantiated. This class MUST NOT require any constructor arguments.
-- ``condition``: The condition that is evaluated to boolean to decide whether the class can be instantiated.
+ When you are working on an HTTP Client or Message Factory, you can easily make your class discoverable:
+ you have to configure it as a Puli resource (`binding`_ in Puli terminology).
 
-The following types are allowed:
-    - string: Checked for class existence
-    - callable: Called and evaluated to boolean
-    - boolean: Can be true or false
-    - any other: considered false
+A binding must have a type, called `binding-type`_. All of our interfaces are registered as binding types.
 
-The condition can also be omitted. In that case the class is used as the condition (to check for existence).
+For example: a client ``Http\Client\MyClient`` should be bind to ``Http\Client\HttpClient``
 
-Classes registered manually are put on top of the list.
+Puli uses a ``puli.json`` file for configuration (placed in the package root).
 
-Writing your own discovery
-^^^^^^^^^^^^^^^^^^^^^^^^^^
+An example for a client's bindings::
 
-Each discovery service is based on the ``ClassDiscovery`` and has to specify a ``cache`` property and a ``class`` property
-to specify classes for the corresponding service. Since they are static, this properties need to be re-declared
-in each discovery class. If ``ClassDiscovery`` would declare them,
-they would be shared between the discovery classes which would make no sense.
-
-Here is an example discovery::
+    {
+        "version": "1.0",
+        "name": "php-http/guzzle6-adapter",
+        "bindings": {
+            "04b5a002-71a8-473d-a8df-75671551b84a": {
```

Comment:
I explain that later:

> Puli provides a CLI tool for configuring bindings. It is necessary, because each binding must have a unique identifier.

Maybe we need to explain it a bit more.